base/special/exp.jl

```diff
@@ -1,65 +1,184 @@
 
-# for values smaller than this threshold just use a Taylor expansion
-@eval exp_small_thres(::Type{Float64}) = $(2.0^-28)
-@eval exp_small_thres(::Type{Float32}) = $(2.0f0^-13)
+@inline function table_unpack(ind)
+    j = @inbounds J_TABLE[ind]
+    jU = reinterpret(Float64, JU_CONST | (j&JU_MASK))
+    jL = reinterpret(Float64, JL_CONST | (j>>8))
+    return jU, jL
+end
 
-"""
+# Method
+# 1. Argument reduction: Reduce x to an r so that |r| <= log(2)/512. Given x, base b,
+#    find r and integers k, j such that
+#       x = (k + j/256)*log(b, 2) + r,  0 <= j < 256, |r| <= log(b,2)/512.
+#
+# 2. Approximate b^r-1 by 3rd-degree minimax polynomial p_b(r) on the interval [-log(b,2)/512, log(b,2)/512].
+#    Since the bounds on r are very tight, this is sufficient to be accurate to floating point epsilon.
+#
+# 3. Scale back: b^x = 2^k * 2^(j/256) * (1 + p_b(r))
+#    Since the range of possible j is small, 2^(j/256) is stored for all possible values in slightly extended precision.
+for (func, base) in (:exp2=>Val(2), :exp=>Val(ℯ), :exp10=>Val(10))
+    @eval begin
+        ($func)(x::Real) = ($func)(float(x))
+        @inline function ($func)(x::T) where T<:Float64
+            N_float = muladd(x, LogBo256INV($base, T), MAGIC_ROUND_CONST(T))
+            N = reinterpret(uinttype(T), N_float) % Int32
+            N_float -=  MAGIC_ROUND_CONST(T)
```

Comment:
I don't understand this `MAGIC_ROUND_CONST`, bit.   Can't you just use `round` (it's one instruction)  and if you need an Int `unsafe_trunc` ?

Reply:
from my tests, for Float64, rounding is slower (largely since you can't fold it with multiplication).

Reply:
Acording to my timings, using the strategy I use for `Float32` for `Float64` has about the same scalar performance, but is 35% slower for simd performance. Given this, would you prefer I make a change to make it use rounding? I agree it would be clearer.

Reply:
Thanks for looking into it. In general, I'm not a big fan of these types of hacks, as you mention, since they can be inscrutable/hide meaning and  aren't as clean. Personally, given the fact that `round`  is a single instruction, even with the slow down for simd perf,  I would opt to make that change. It's possible that the instruction might be faster on different arch's in the future and with compiler upgrades, so the performance difference could be up for grabs. I'll leave the decision up to.

Reply:
Maybe next to the definition of the "magic round constants", explain how it works?
```julia
julia> using BetterExp # https://github.com/oscardssmith/BetterExp.jl

julia> eps(BetterExp.MAGIC_ROUND_CONST(Float64))
 1.0

julia> eps(BetterExp.MAGIC_ROUND_CONST(Float32))
 1.0f0
```
I.e., that adding forces the sum to be rounded to the nearest integer (for `x` in the valid range of `exp`; for `abs(x)` larger than the magic constant, we'll get `0.0` or `Inf` anyway).
Most programmers are already going to be at least passingly familiar with floating point rounding.

And a comment next to `N_float -= MAGIC_ROUND_CONST(T)` saying `N_float` now equals rounded `x * LogBo256INV($base, T)`?

35% performance is a lot to sacrifice.
But that relative speed of these instructions is architecture dependent is a valid observation.
Looking at [Agner Fog's instruction tables](https://www.agner.org/optimize/instruction_tables.pdf), the latency (reciprocal throughput) figures are:

| CPU | SIMD Round | SIMD add/subtract |
|---|---|---|
| Zen2 | 3 (1) | 3 (0.5)
| Skylake-X | 8 (1) | 4 (0.5) |
| Skylake | 8 (1) | 4 (0.5) |
| Haswell | 6 (2) | 3 (1) |

All these CPUs have twice the arithmetic throughput, but for Zen2 the latency of rounding and add/subtract are equal.
This means, in Zen it would take the same number of clock cycles (3) after starting `round(N_float)` before the result can be used for a subsequent calculation as it would after starting `N_float -= MAGIC_ROUND_CONST(T)`.
But Haswell and Skylake require 3 or 4 extra clock cycles.

If I understand correctly, it also looks like rounding uses 2 execution ports (e.g., "2p01/05") on the Intel CPUs, while most instructions like arithmetic require only 1, which is why the reciprocal throughput is twice as high.
However, on Zen2 it requires just a single port, but there's only 1 port available.
Hence, rounding should mix better with other instructions on Zen2 than it should on most recent Intel.

So it'd be interesting if someone with a Zen2 CPU (e.g., 3700X) could benchmark both versions.

Reply:
my cpu is a 3600.

Reply:
I think the reason rounding is still so much slower for simd is that the floating point subtraction time can be masked by the table lookup.

Reply:
I'd be happy to change it to a round instruction to get it merged, and we can decide later if we care about the extra speedup.

Reply:
> I.e., that adding forces the sum to be rounded to the nearest integer (for x in the valid range of exp; for abs(x) larger than the magic constant, we'll get 0.0 or Inf anyway).

Sounds resonable. 

I'm curious why compilers can't already optimize for this. Seems like an obvious thing to do.

Reply:
My guess is at least part of the problem is that this would do very different things for large/small numbers, Infs, Nans. All these cases end up with the same behavior in the end, but I can see it being hard for the compiler to realize this.

Reply:
@oscardssmith I think there's a misunderstanding, I was agreeing with @chriselrod to perhaps simply add a basic comment regarding:

> Maybe next to the definition of the "magic round constants", explain how it works?

> 
> julia> using BetterExp # https://github.com/oscardssmith/BetterExp.jl
> 
> julia> eps(BetterExp.MAGIC_ROUND_CONST(Float64))
>  1.0
> 
> julia> eps(BetterExp.MAGIC_ROUND_CONST(Float32))
>  1.0f0
> I.e., that adding forces the sum to be rounded to the nearest integer (for x in the valid range of exp; for abs(x) larger than the magic constant, we'll get 0.0 or Inf anyway).
> Most programmers are already going to be at least passingly familiar with floating point rounding.
> 
> And a comment next to N_float -= MAGIC_ROUND_CONST(T) saying N_float now equals rounded x * LogBo256INV($base, T)?


Yeah, I'm not a fan of this type of hack, but since it seems like it's too much performance to leave on the table according to @chriselrod, perhaps it's tolerable.  Any other thoughts @chriselrod ?

Reply:
yep. I misunderstood. magic has been re-added with an explanation

Reply:
> my cpu is a 3600.

Oh, I should've remembered that.

> Sounds resonable.

@oscardssmith 
I would prefer the magic const version given that performance difference; mind using it and adding an explanation of how it works?

> I'm curious why compilers can't already optimize for this. Seems like an obvious thing to do.

I think the problem is that it's limited in applicability.
Rounding was generally twice as slow as the arithmetic, so replacing a round with an add and a subtract doesn't help.

But each of them that we fuse with a multiplication becomes more or less free depending on the relative cost of fmas and add/subtract*, so whether or not it helps depends on these fusion-opportunities. 

* That cost is variable. On Skylake(-X), they're equivalent. On Haswell, fma is actually faster than add/subtract (meaning `1.0 * a + b` is faster than `a + b` on Haswell, if you happen to already have a `1.0` laying around...). On Zen(2), fma is slower (but still much faster than separate multiplication and addition/subtraction). Obviously no benefits on CPUs without fma.

Another problem is that optimization is only valid in part of `x`'s range. It could accidentally yield `Inf` from finite inputs, for example.
The valid range of that trick is a superset of the valid range of `exp`'s inputs, so it's fine here, but not as an arbitrarily applied optimization.

Finally, if we look at the code
```julia
            N_float = muladd(x, LogBo256INV($base, T), MAGIC_ROUND_CONST(T))
            N = reinterpret(uinttype(T), N_float) % Int32
            N_float -=  MAGIC_ROUND_CONST(T)
```
We need to `N = reinterpret(uinttype(T), muladd(x, LogBo256INV($base, T), MAGIC_ROUND_CONST(T)))` here. All the other changes needed to get the correct `k` add up.